# Hoja de Trabajo \# 4

---


por Josué Obregón <br>
DS6011 - Feature Engineering <br>
UVG Masters - Escuela de Negocios<br>


## Objetivos

El objetivo de esta hoja de trabajo  es presentar al estudiante diferentes técnicas de codificación de variables categóricas. Éstas técnicas incluyen codificadores clásicos, de contraste y codificadores supervisados o bayesianos.

También se busca que el estudiante practique la utilización de éstas técnicas con las librerías disponibles en el lenguaje Python.


## Importación de librerías y carga de los datos a varios pandas [DataFrames](https://pandas.pydata.org/pandas-docs/version/1.1.5/reference/frame.html)




In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

In [2]:
!mkdir data

In [3]:
import gdown

urls = ['https://drive.google.com/uc?export=download&id=16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq', # okc_train  https://drive.google.com/file/d/16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq/view?usp=sharing
        ]
outputs = ['okc_train.csv']
for url,output in zip(urls,outputs):
  gdown.download(url, f'data/{output}', quiet=False)

Downloading...
From: https://drive.google.com/uc?export=download&id=16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq
To: /content/data/okc_train.csv
12.3MB [00:00, 80.5MB/s]


In [4]:
df = pd.read_csv('data/okc_train.csv',index_col=0)

In [5]:
df.head()

age               diet      drinks  ... essay_length  Class  profile
1    22  strictly_anything    socially  ...     3.420945  other        1
4    23         vegetarian    socially  ...     2.671173  other        3
6    29    mostly_anything    socially  ...     3.391112   stem        5
8    31    mostly_anything    socially  ...     3.090611  other        6
10   37    mostly_anything  not_at_all  ...     3.382377  other        7

[5 rows x 38 columns]

In [6]:
df['drinks'].value_counts()

socially          27521
rarely             3981
often              3413
not_at_all         2157
drinks_missing     1208
very_often          319
desperately         210
Name: drinks, dtype: int64

In [7]:
df['status'].value_counts()

single            36014
seeing_someone     1354
available          1234
married             204
unknown               3
Name: status, dtype: int64

Pequeño dataset de prueba para algunas explicaciones y para las figuras en las diapositivas


In [8]:
df_test = pd.DataFrame({
... 'City': ['SF', 'SF', 'SF', 'NYC', 'NYC', 'NYC',
... 'Seattle', 'Seattle', 'Seattle'],
... 'Rent': [3999, 4000, 4001, 3499, 3500, 3501, 2499, 2500, 2501]
... })

In [9]:
df_test

City  Rent
0       SF  3999
1       SF  4000
2       SF  4001
3      NYC  3499
4      NYC  3500
5      NYC  3501
6  Seattle  2499
7  Seattle  2500
8  Seattle  2501

Para esta hoja de trabajo vamos a utilizar la librería [category_encoders](http://contrib.scikit-learn.org/category_encoders/index.html), la cual es compatible con scikit-learn.

In [10]:
!pip install category_encoders

     |████████████████████████████████| 81kB 6.6MB/s 


In [11]:
col = 'drinks'

# Unsupervised Encoders

## Classic Encoders

### One-Hot Encoder

In [12]:
from category_encoders import OneHotEncoder

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning:

pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.



In [13]:
onehot_enc = OneHotEncoder(use_cat_names=True)

In [14]:
onehot_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks_socially  ...  drinks_desperately
1                    1  ...                   0
4                    1  ...                   0
6                    1  ...                   0
8                    1  ...                   0
10                   0  ...                   0
...                ...  ...                 ...
59940                1  ...                   0
59943                0  ...                   0
59944                0  ...                   0
59945                1  ...                   0
59946                1  ...                   0

[38809 rows x 7 columns]

In [15]:
onehot_enc.mapping[0]['mapping']

drinks_socially  drinks_not_at_all  ...  drinks_very_often  drinks_desperately
 1                1                  0  ...                  0                   0
 2                0                  1  ...                  0                   0
 3                0                  0  ...                  0                   0
 4                0                  0  ...                  0                   0
 5                0                  0  ...                  0                   0
 6                0                  0  ...                  1                   0
 7                0                  0  ...                  0                   1
-1                0                  0  ...                  0                   0
-2                0                  0  ...                  0                   0

[9 rows x 7 columns]

### Binary Encoder

In [16]:
from category_encoders import BinaryEncoder

In [17]:
binary_enc = BinaryEncoder()

In [18]:
binary_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks_0  drinks_1  drinks_2  drinks_3
1             0         0         0         1
4             0         0         0         1
6             0         0         0         1
8             0         0         0         1
10            0         0         1         0
...         ...       ...       ...       ...
59940         0         0         0         1
59943         0         1         0         0
59944         0         0         1         0
59945         0         0         0         1
59946         0         0         0         1

[38809 rows x 4 columns]

In [19]:
binary_enc.base_n_encoder.mapping[0]['mapping']

drinks_0  drinks_1  drinks_2  drinks_3
 1         0         0         0         1
 2         0         0         1         0
 3         0         0         1         1
 4         0         1         0         0
 5         0         1         0         1
 6         0         1         1         0
 7         0         1         1         1
-1         0         0         0         0
-2         0         0         0         0

### Frequency or Count Encoder

In [20]:
from category_encoders import CountEncoder

In [21]:
count_enc = CountEncoder(normalize=False)

In [22]:
df[col].value_counts()

socially          27521
rarely             3981
often              3413
not_at_all         2157
drinks_missing     1208
very_often          319
desperately         210
Name: drinks, dtype: int64

In [50]:
count_enc.fit_transform(df[col])

drinks
1      0.709140
4      0.709140
6      0.709140
8      0.709140
10     0.055580
...         ...
59940  0.709140
59943  0.087944
59944  0.055580
59945  0.709140
59946  0.709140

[38809 rows x 1 columns]

In [51]:
count_enc.mapping


{'drinks': socially          0.709140
 rarely            0.102579
 often             0.087944
 not_at_all        0.055580
 drinks_missing    0.031127
 otros             0.013631
 Name: drinks, dtype: float64}

Intentemos utilizando las funciones de combinacion de grupos

In [52]:
count_enc = CountEncoder(combine_min_nan_groups=True, min_group_size=500, min_group_name=
'Otros')


In [53]:
count_enc.fit_transform(df[col])


drinks
1       27521
4       27521
6       27521
8       27521
10       2157
...       ...
59940   27521
59943    3413
59944    2157
59945   27521
59946   27521

[38809 rows x 1 columns]

In [54]:
count_enc.mapping


{'drinks': socially          27521
 rarely             3981
 often              3413
 not_at_all         2157
 drinks_missing     1208
 Otros               529
 Name: drinks, dtype: int64}

### Ordinal Encoder

In [55]:
from category_encoders import OrdinalEncoder

In [56]:
ord_enc = OrdinalEncoder()

In [26]:
ord_enc.fit_transform(df[col])

drinks
1           1
4           1
6           1
8           1
10          2
...       ...
59940       1
59943       4
59944       2
59945       1
59946       1

[38809 rows x 1 columns]

In [57]:
drinks_dict= {None: 0, 'not_at_all':1,'rarelly':2, 'socially':3, 'often':4,'very_often':
5,'desperately':6,'drinks_missing':-1,}
col_drinks_dict = {'col':'drinks', 'mapping':drinks_dict}
ordinal_enc = OrdinalEncoder(mapping=[col_drinks_dict])


In [58]:
ordinal_enc.fit_transform(df[col])


drinks
1         3.0
4         3.0
6         3.0
8         3.0
10        1.0
...       ...
59940     3.0
59943     4.0
59944     1.0
59945     3.0
59946     3.0

[38809 rows x 1 columns]

In [59]:
ordinal_enc.category_mapping[0]['mapping']


{None: 0,
 'desperately': 6,
 'drinks_missing': -1,
 'not_at_all': 1,
 'often': 4,
 'rarelly': 2,
 'socially': 3,
 'very_often': 5}

###Feature Hashing

In [29]:
from sklearn.feature_extraction import FeatureHasher


In [60]:
hash_enc = FeatureHasher(n_features=8, input_type='string', alternate_sign=True)


In [61]:
col='where_town'
hashed_features=hash_enc.fit_transform(df[col])


In [62]:
hashed_features.toarray()


array([[ 1., -1.,  2., ..., -2.,  0., -3.],
       [ 0.,  1.,  0., ..., -1.,  1.,  2.],
       [ 0., -1.,  2., ..., -1.,  0., -2.],
       ...,
       [ 1., -1.,  2., ..., -2.,  0., -3.],
       [ 0., -1.,  2., ..., -1.,  0., -2.],
       [ 0., -1.,  2., ..., -1.,  0., -2.]])

In [63]:
np_hashed = np.array(hashed_features.todense())
np.unique(np_hashed, axis=0)


array([[ 0., -2.,  2.,  1.,  0., -2., -1.,  0.],
       [ 0., -1.,  0.,  4.,  0., -1.,  1.,  1.],
       [ 0., -1.,  1.,  0., -1., -3.,  0.,  1.],
       [ 0., -1.,  1.,  1.,  1., -1.,  2.,  0.],
       [ 0., -1.,  2.,  3., -2., -1.,  0., -2.],
       [ 0.,  0.,  0.,  0.,  1., -1.,  0.,  1.],
       [ 0.,  0.,  0.,  1.,  0., -1., -1.,  1.],
       [ 0.,  0.,  0.,  2.,  0., -2.,  0.,  0.],
       [ 0.,  0.,  1., -1.,  0., -1.,  0.,  1.],
       [ 0.,  0.,  1.,  0., -1.,  0.,  0.,  0.],
       [ 0.,  0.,  1.,  0.,  2.,  0.,  0.,  0.],
       [ 0.,  0.,  1.,  0.,  5., -1.,  1.,  1.],
       [ 0.,  0.,  1.,  3., -1.,  0.,  0.,  0.],
       [ 0.,  0.,  2., -1.,  1.,  0.,  1.,  2.],
       [ 0.,  0.,  2.,  0.,  0., -1.,  1.,  0.],
       [ 0.,  0.,  2.,  0.,  2., -2., -1.,  0.],
       [ 0.,  0.,  2.,  0.,  2.,  0.,  1.,  0.],
       [ 0.,  0.,  2.,  1.,  0., -1.,  1.,  0.],
       [ 0.,  0.,  2.,  1.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  2.,  2.,  0., -1., -1., -1.],
       [ 0.,  0.,  2

In [64]:
np.unique(np_hashed, axis=0).shape

(50, 8)

In [35]:
np_hashed = np.array(hashed_features.todense())

In [36]:
np.unique(np_hashed, axis=0)

array([[ 0., -2.,  2.,  1.,  0., -2., -1.,  0.],
       [ 0., -1.,  0.,  4.,  0., -1.,  1.,  1.],
       [ 0., -1.,  1.,  0., -1., -3.,  0.,  1.],
       [ 0., -1.,  1.,  1.,  1., -1.,  2.,  0.],
       [ 0., -1.,  2.,  3., -2., -1.,  0., -2.],
       [ 0.,  0.,  0.,  0.,  1., -1.,  0.,  1.],
       [ 0.,  0.,  0.,  1.,  0., -1., -1.,  1.],
       [ 0.,  0.,  0.,  2.,  0., -2.,  0.,  0.],
       [ 0.,  0.,  1., -1.,  0., -1.,  0.,  1.],
       [ 0.,  0.,  1.,  0., -1.,  0.,  0.,  0.],
       [ 0.,  0.,  1.,  0.,  2.,  0.,  0.,  0.],
       [ 0.,  0.,  1.,  0.,  5., -1.,  1.,  1.],
       [ 0.,  0.,  1.,  3., -1.,  0.,  0.,  0.],
       [ 0.,  0.,  2., -1.,  1.,  0.,  1.,  2.],
       [ 0.,  0.,  2.,  0.,  0., -1.,  1.,  0.],
       [ 0.,  0.,  2.,  0.,  2., -2., -1.,  0.],
       [ 0.,  0.,  2.,  0.,  2.,  0.,  1.,  0.],
       [ 0.,  0.,  2.,  1.,  0., -1.,  1.,  0.],
       [ 0.,  0.,  2.,  1.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  2.,  2.,  0., -1., -1., -1.],
       [ 0.,  0.,  2

In [65]:
np.unique(np_hashed,axis=0).shape

(50, 8)

In [66]:
from sys import getsizeof

In [39]:
print('Our pandas Series, in bytes: ', getsizeof(df[col]))
print('Our hashed numpy array, in bytes: ', getsizeof(hashed_features))

Our pandas Series, in bytes:  2944494
Our hashed numpy array, in bytes:  64


## Contrast Encoders

### Diferencia entre One-hot Encoding y Dummy Coding

In [67]:
df_test

City  Rent
0       SF  3999
1       SF  4000
2       SF  4001
3      NYC  3499
4      NYC  3500
5      NYC  3501
6  Seattle  2499
7  Seattle  2500
8  Seattle  2501

In [68]:
print('Global mean: ',df_test['Rent'].mean())
print(df_test.groupby('City').mean())


Global mean:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


In [69]:
one_hot_df = pd.get_dummies(df_test, prefix=['city']) #

In [70]:
one_hot_df


Rent  city_NYC  city_SF  city_Seattle
0  3999         0        1             0
1  4000         0        1             0
2  4001         0        1             0
3  3499         1        0             0
4  3500         1        0             0
5  3501         1        0             0
6  2499         0        0             1
7  2500         0        0             1
8  2501         0        0             1

Con la codificación one-hot, el término de intersección (intercepto) representa la media global de la variable objetivo: 'Renta', y cada uno de los coeficientes lineales representa que tanto difiere la renta media de esa ciudad respecto a la media global.

In [71]:
from sklearn import linear_model


In [72]:
model = linear_model.LinearRegression()
model.fit(one_hot_df[['city_NYC','city_SF','city_Seattle']], one_hot_df['Rent'])


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [73]:
print(f'Coefficients: {model.coef_}')
print(f'Intercept: {model.intercept_}') 

Coefficients: [ 166.66666667  666.66666667 -833.33333333]
Intercept: 3333.3333333333335


Ahora entrenemos el mismo modelo con dummy coding

In [74]:
dummy_df = pd.get_dummies(df_test, prefix=['city'], drop_first=True)


In [75]:
dummy_df

Rent  city_SF  city_Seattle
0  3999        1             0
1  4000        1             0
2  4001        1             0
3  3499        0             0
4  3500        0             0
5  3501        0             0
6  2499        0             1
7  2500        0             1
8  2501        0             1

Con codificación dummy, el coeficiente de sesgo (intercepto) representa el valor medio de la variable $y$ para la categoría de referencia, que en el ejemplo es la ciudad NYC. El coeficiente para la *i*-ésima característica es igual a la diferencia entre el valor medio de la respuesta para la i-ésima categoría y la media de la categoría de referencia.

In [76]:
model.fit(dummy_df[['city_SF','city_Seattle']], dummy_df['Rent'])
print(f'Coefficients: {model.coef_}')
print(f'Intercept: {model.intercept_}')


Coefficients: [  500. -1000.]
Intercept: 3500.0


In [77]:
print('Global mean: ',df_test['Rent'].mean())
print(df_test.groupby('City').mean())


Global mean:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


### Dummy Coding

In [40]:
col = 'drinks'

In [41]:
from sklearn.preprocessing import OneHotEncoder as OneHotEncoder_sk 

In [42]:
dummy_enc = OneHotEncoder_sk(drop='first',sparse=False)

In [43]:
dummy_enc.fit_transform(df[[col]])

array([[0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.],
       ...,
       [0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0.]])

In [44]:
dummy_enc.categories_

[array(['desperately', 'drinks_missing', 'not_at_all', 'often', 'rarely',
        'socially', 'very_often'], dtype=object)]

In [45]:
dummy_df2 = pd.get_dummies(df[col], prefix='dr', drop_first=True)
dummy_df2.head()

dr_drinks_missing  dr_not_at_all  ...  dr_socially  dr_very_often
1                   0              0  ...            1              0
4                   0              0  ...            1              0
6                   0              0  ...            1              0
8                   0              0  ...            1              0
10                  0              1  ...            0              0

[5 rows x 6 columns]

In [46]:
dummy_df2.drop_duplicates()

dr_drinks_missing  dr_not_at_all  ...  dr_socially  dr_very_often
1                    0              0  ...            1              0
10                   0              1  ...            0              0
19                   0              0  ...            0              0
41                   0              0  ...            0              0
93                   1              0  ...            0              0
97                   0              0  ...            0              1
549                  0              0  ...            0              0

[7 rows x 6 columns]


### Sum (or Deviation) Coding

In [47]:
from category_encoders import SumEncoder

In [48]:
sum_enc = SumEncoder()

In [78]:
sum_enc.fit_transform(df[[col]])


/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  where_town_0  ...  where_town_48  where_town_49
1              1           1.0  ...            0.0            0.0
4              1           0.0  ...            0.0            0.0
6              1           0.0  ...            0.0            0.0
8              1           0.0  ...            0.0            0.0
10             1           0.0  ...            0.0            0.0
...          ...           ...  ...            ...            ...
59940          1           0.0  ...            0.0            0.0
59943          1           0.0  ...            0.0            0.0
59944          1           1.0  ...            0.0            0.0
59945          1           0.0  ...            0.0            0.0
59946          1           0.0  ...            0.0            0.0

[38809 rows x 51 columns]

In [79]:
sum_enc.mapping[0]['mapping']


where_town_0  where_town_1  ...  where_town_48  where_town_49
 1            1.0           0.0  ...            0.0            0.0
 2            0.0           1.0  ...            0.0            0.0
 3            0.0           0.0  ...            0.0            0.0
 4            0.0           0.0  ...            0.0            0.0
 5            0.0           0.0  ...            0.0            0.0
 6            0.0           0.0  ...            0.0            0.0
 7            0.0           0.0  ...            0.0            0.0
 8            0.0           0.0  ...            0.0            0.0
 9            0.0           0.0  ...            0.0            0.0
 10           0.0           0.0  ...            0.0            0.0
 11           0.0           0.0  ...            0.0            0.0
 12           0.0           0.0  ...            0.0            0.0
 13           0.0           0.0  ...            0.0            0.0
 14           0.0           0.0  ...            0.0            0.0
 15           0.0           0.0  ...            0.0            0.0
 16           0.0           0.0  ...            0.0            0.0
 17           0.0           0.0  ...            0.0            0.0
 18           0.0           0.0  ...            0.0            0.0
 19           0.0           0.0  ...            0.0            0.0
 20           0.0           0.0  ...            0.0            0.0
 21           0.0           0.0  ...            0.0            0.0
 22           0.0           0.0  ...            0.0            0.0
 23           0.0           0.0  ...            0.0            0.0
 24           0.0           0.0  ...            0.0            0.0
 25           0.0           0.0  ...            0.0            0.0
 26           0.0           0.0  ...            0.0            0.0
 27           0.0           0.0  ...            0.0            0.0
 28           0.0           0.0  ...            0.0            0.0
 29           0.0           0.0  ...            0.0            0.0
 30           0.0           0.0  ...            0.0            0.0
 31           0.0           0.0  ...            0.0            0.0
 32           0.0           0.0  ...            0.0            0.0
 33           0.0           0.0  ...            0.0            0.0
 34           0.0           0.0  ...            0.0            0.0
 35           0.0           0.0  ...            0.0            0.0
 36           0.0           0.0  ...            0.0            0.0
 37           0.0           0.0  ...            0.0            0.0
 38           0.0           0.0  ...            0.0            0.0
 39           0.0           0.0  ...            0.0            0.0
 40           0.0           0.0  ...            0.0            0.0
 41           0.0           0.0  ...            0.0            0.0
 42           0.0           0.0  ...            0.0            0.0
 43           0.0           0.0  ...            0.0            0.0
 44           0.0           0.0  ...            0.0            0.0
 45           0.0           0.0  ...            0.0            0.0
 46           0.0           0.0  ...            0.0            0.0
 47           0.0           0.0  ...            0.0            0.0
 48           0.0           0.0  ...            0.0            0.0
 49           0.0           0.0  ...            1.0            0.0
 50           0.0           0.0  ...            0.0            1.0
 51          -1.0          -1.0  ...           -1.0           -1.0
-1            0.0           0.0  ...            0.0            0.0
-2            0.0           0.0  ...            0.0            0.0

[53 rows x 50 columns]

Utilizando el ejemplo de prueba

In [80]:
sum_enc = SumEncoder()
sum_df = sum_enc.fit_transform(df_test['City'])
sum_df


/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  City_0  City_1
0          1     1.0     0.0
1          1     1.0     0.0
2          1     1.0     0.0
3          1     0.0     1.0
4          1     0.0     1.0
5          1     0.0     1.0
6          1    -1.0    -1.0
7          1    -1.0    -1.0
8          1    -1.0    -1.0

In [81]:
sum_enc.mapping[0]['mapping']


City_0  City_1
 1     1.0     0.0
 2     0.0     1.0
 3    -1.0    -1.0
-1     0.0     0.0
-2     0.0     0.0

In [82]:
sum_enc.ordinal_encoder.category_mapping[0]['mapping']


SF         1
NYC        2
Seattle    3
NaN       -2
dtype: int64

In [83]:
sum_df['Rent']=df_test['Rent']
sum_df


intercept  City_0  City_1  Rent
0          1     1.0     0.0  3999
1          1     1.0     0.0  4000
2          1     1.0     0.0  4001
3          1     0.0     1.0  3499
4          1     0.0     1.0  3500
5          1     0.0     1.0  3501
6          1    -1.0    -1.0  2499
7          1    -1.0    -1.0  2500
8          1    -1.0    -1.0  2501

In [84]:
model = linear_model.LinearRegression()
model.fit(sum_df[['City_0','City_1']], sum_df['Rent'])



LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

La codificación de efectos (sum coding) es muy similar a la codificación dummy, pero da como resultado modelos de regresión lineal que son aún más simples de interpretar.

En el ejemplo, vemos que el término de intersección representa la media global del la variable respuesta, y los coeficientes individuales indican cuánto las medias de las categorías individuales difieren de la media global. (Esto se llama el efecto principal de la categoría o nivel, de ahí el nombre "codificación de efectos").

La codificación One-hot, encontró la misma intersección y coeficientes, pero en ese caso hay coeficientes lineales para cada ciudad. En la codificación de efecto, ningun feature representa la categoría de referencia, ***por lo que el efecto de la categoría de referencia debe calcularse por separado como la suma negativa de los coeficientes de todas las demás categorías.***

In [ ]:
print(f'Coefficients: {model.coef_}')
print(f'Intercept: {model.intercept_}')
print(f'Negative sum of coaefficients:{np.sum(model.coef_*-1)}')


In [ ]:
print('Global mean: ', df_test['Rent'].mean())
print(df_test.groupby('City').mean())

Restulados de la regresión lineal con One-hot encoding

Coefficients: $[ 166.66666667,   666.66666667,  -833.33333333]$

Intercept: $3333.3333333333335$

### Backward Difference Coding

In [85]:
from category_encoders import BackwardDifferenceEncoder

In [86]:
back_diff_enc = BackwardDifferenceEncoder()

In [87]:
back_diff_enc.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  where_town_0  ...  where_town_48  where_town_49
1              1     -0.980392  ...      -0.039216      -0.019608
4              1      0.019608  ...      -0.039216      -0.019608
6              1      0.019608  ...      -0.039216      -0.019608
8              1      0.019608  ...      -0.039216      -0.019608
10             1      0.019608  ...      -0.039216      -0.019608
...          ...           ...  ...            ...            ...
59940          1      0.019608  ...      -0.039216      -0.019608
59943          1      0.019608  ...      -0.039216      -0.019608
59944          1     -0.980392  ...      -0.039216      -0.019608
59945          1      0.019608  ...      -0.039216      -0.019608
59946          1      0.019608  ...      -0.039216      -0.019608

[38809 rows x 51 columns]

In [88]:
back_diff_enc.mapping[0]['mapping']

where_town_0  where_town_1  ...  where_town_48  where_town_49
 1      -0.980392     -0.960784  ...      -0.039216      -0.019608
 2       0.019608     -0.960784  ...      -0.039216      -0.019608
 3       0.019608      0.039216  ...      -0.039216      -0.019608
 4       0.019608      0.039216  ...      -0.039216      -0.019608
 5       0.019608      0.039216  ...      -0.039216      -0.019608
 6       0.019608      0.039216  ...      -0.039216      -0.019608
 7       0.019608      0.039216  ...      -0.039216      -0.019608
 8       0.019608      0.039216  ...      -0.039216      -0.019608
 9       0.019608      0.039216  ...      -0.039216      -0.019608
 10      0.019608      0.039216  ...      -0.039216      -0.019608
 11      0.019608      0.039216  ...      -0.039216      -0.019608
 12      0.019608      0.039216  ...      -0.039216      -0.019608
 13      0.019608      0.039216  ...      -0.039216      -0.019608
 14      0.019608      0.039216  ...      -0.039216      -0.019608
 15      0.019608      0.039216  ...      -0.039216      -0.019608
 16      0.019608      0.039216  ...      -0.039216      -0.019608
 17      0.019608      0.039216  ...      -0.039216      -0.019608
 18      0.019608      0.039216  ...      -0.039216      -0.019608
 19      0.019608      0.039216  ...      -0.039216      -0.019608
 20      0.019608      0.039216  ...      -0.039216      -0.019608
 21      0.019608      0.039216  ...      -0.039216      -0.019608
 22      0.019608      0.039216  ...      -0.039216      -0.019608
 23      0.019608      0.039216  ...      -0.039216      -0.019608
 24      0.019608      0.039216  ...      -0.039216      -0.019608
 25      0.019608      0.039216  ...      -0.039216      -0.019608
 26      0.019608      0.039216  ...      -0.039216      -0.019608
 27      0.019608      0.039216  ...      -0.039216      -0.019608
 28      0.019608      0.039216  ...      -0.039216      -0.019608
 29      0.019608      0.039216  ...      -0.039216      -0.019608
 30      0.019608      0.039216  ...      -0.039216      -0.019608
 31      0.019608      0.039216  ...      -0.039216      -0.019608
 32      0.019608      0.039216  ...      -0.039216      -0.019608
 33      0.019608      0.039216  ...      -0.039216      -0.019608
 34      0.019608      0.039216  ...      -0.039216      -0.019608
 35      0.019608      0.039216  ...      -0.039216      -0.019608
 36      0.019608      0.039216  ...      -0.039216      -0.019608
 37      0.019608      0.039216  ...      -0.039216      -0.019608
 38      0.019608      0.039216  ...      -0.039216      -0.019608
 39      0.019608      0.039216  ...      -0.039216      -0.019608
 40      0.019608      0.039216  ...      -0.039216      -0.019608
 41      0.019608      0.039216  ...      -0.039216      -0.019608
 42      0.019608      0.039216  ...      -0.039216      -0.019608
 43      0.019608      0.039216  ...      -0.039216      -0.019608
 44      0.019608      0.039216  ...      -0.039216      -0.019608
 45      0.019608      0.039216  ...      -0.039216      -0.019608
 46      0.019608      0.039216  ...      -0.039216      -0.019608
 47      0.019608      0.039216  ...      -0.039216      -0.019608
 48      0.019608      0.039216  ...      -0.039216      -0.019608
 49      0.019608      0.039216  ...      -0.039216      -0.019608
 50      0.019608      0.039216  ...       0.960784      -0.019608
 51      0.019608      0.039216  ...       0.960784       0.980392
-1       0.000000      0.000000  ...       0.000000       0.000000
-2       0.000000      0.000000  ...       0.000000       0.000000

[53 rows x 50 columns]

Utilizando el ejemplo de prueba

In [89]:
back_dif_enc = BackwardDifferenceEncoder()


In [91]:
back_dif_df = back_dif_enc.fit_transform(df_test['City'])
back_dif_df['Rent']=df_test['Rent']
back_dif_df


/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept    City_0    City_1  Rent
0          1 -0.666667 -0.333333  3999
1          1 -0.666667 -0.333333  4000
2          1 -0.666667 -0.333333  4001
3          1  0.333333 -0.333333  3499
4          1  0.333333 -0.333333  3500
5          1  0.333333 -0.333333  3501
6          1  0.333333  0.666667  2499
7          1  0.333333  0.666667  2500
8          1  0.333333  0.666667  2501

In [92]:
back_dif_enc.mapping[0]['mapping']


City_0    City_1
 1 -0.666667 -0.333333
 2  0.333333 -0.333333
 3  0.333333  0.666667
-1  0.000000  0.000000
-2  0.000000  0.000000

In [93]:
back_dif_enc.ordinal_encoder.category_mapping[0]['mapping']


SF         1
NYC        2
Seattle    3
NaN       -2
dtype: int64

In [94]:
model = linear_model.LinearRegression()
model.fit(back_dif_df[['City_0','City_1']], back_dif_df['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

La codificación de diferencia en reversa es útil para codificar variables ordinales.

En el ejemplo, vemos que el término de intersección representa la media global de la variable respuesta, y los coeficientes individuales indican cuánto las medias de las categorías individuales difieren de la media de la categoría inmediatamente anterior.

In [96]:
print(f'Coefficients: {model.coef_}')
print(f'Intercept: {model.intercept_}')

Coefficients: [ -500. -1000.]
Intercept: 3333.3333333333335


In [95]:
print('Global mean: ',df_test['Rent'].mean())
print(df_test.groupby('City').mean())

Global mean:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


### Helmert Coding

In [97]:
from category_encoders import HelmertEncoder
helmer_enc = HelmertEncoder()


In [98]:
helmer_enc.fit_transform(df[[col]])


/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  where_town_0  ...  where_town_48  where_town_49
1              1          -1.0  ...           -1.0           -1.0
4              1           1.0  ...           -1.0           -1.0
6              1           0.0  ...           -1.0           -1.0
8              1           0.0  ...           -1.0           -1.0
10             1           0.0  ...           -1.0           -1.0
...          ...           ...  ...            ...            ...
59940          1           0.0  ...           -1.0           -1.0
59943          1           0.0  ...           -1.0           -1.0
59944          1          -1.0  ...           -1.0           -1.0
59945          1           0.0  ...           -1.0           -1.0
59946          1           0.0  ...           -1.0           -1.0

[38809 rows x 51 columns]

In [99]:
helmer_enc.mapping[0]['mapping']


where_town_0  where_town_1  ...  where_town_48  where_town_49
 1           -1.0          -1.0  ...           -1.0           -1.0
 2            1.0          -1.0  ...           -1.0           -1.0
 3            0.0           2.0  ...           -1.0           -1.0
 4            0.0           0.0  ...           -1.0           -1.0
 5            0.0           0.0  ...           -1.0           -1.0
 6            0.0           0.0  ...           -1.0           -1.0
 7            0.0           0.0  ...           -1.0           -1.0
 8            0.0           0.0  ...           -1.0           -1.0
 9            0.0           0.0  ...           -1.0           -1.0
 10           0.0           0.0  ...           -1.0           -1.0
 11           0.0           0.0  ...           -1.0           -1.0
 12           0.0           0.0  ...           -1.0           -1.0
 13           0.0           0.0  ...           -1.0           -1.0
 14           0.0           0.0  ...           -1.0           -1.0
 15           0.0           0.0  ...           -1.0           -1.0
 16           0.0           0.0  ...           -1.0           -1.0
 17           0.0           0.0  ...           -1.0           -1.0
 18           0.0           0.0  ...           -1.0           -1.0
 19           0.0           0.0  ...           -1.0           -1.0
 20           0.0           0.0  ...           -1.0           -1.0
 21           0.0           0.0  ...           -1.0           -1.0
 22           0.0           0.0  ...           -1.0           -1.0
 23           0.0           0.0  ...           -1.0           -1.0
 24           0.0           0.0  ...           -1.0           -1.0
 25           0.0           0.0  ...           -1.0           -1.0
 26           0.0           0.0  ...           -1.0           -1.0
 27           0.0           0.0  ...           -1.0           -1.0
 28           0.0           0.0  ...           -1.0           -1.0
 29           0.0           0.0  ...           -1.0           -1.0
 30           0.0           0.0  ...           -1.0           -1.0
 31           0.0           0.0  ...           -1.0           -1.0
 32           0.0           0.0  ...           -1.0           -1.0
 33           0.0           0.0  ...           -1.0           -1.0
 34           0.0           0.0  ...           -1.0           -1.0
 35           0.0           0.0  ...           -1.0           -1.0
 36           0.0           0.0  ...           -1.0           -1.0
 37           0.0           0.0  ...           -1.0           -1.0
 38           0.0           0.0  ...           -1.0           -1.0
 39           0.0           0.0  ...           -1.0           -1.0
 40           0.0           0.0  ...           -1.0           -1.0
 41           0.0           0.0  ...           -1.0           -1.0
 42           0.0           0.0  ...           -1.0           -1.0
 43           0.0           0.0  ...           -1.0           -1.0
 44           0.0           0.0  ...           -1.0           -1.0
 45           0.0           0.0  ...           -1.0           -1.0
 46           0.0           0.0  ...           -1.0           -1.0
 47           0.0           0.0  ...           -1.0           -1.0
 48           0.0           0.0  ...           -1.0           -1.0
 49           0.0           0.0  ...           -1.0           -1.0
 50           0.0           0.0  ...           49.0           -1.0
 51           0.0           0.0  ...            0.0           50.0
-1            0.0           0.0  ...            0.0            0.0
-2            0.0           0.0  ...            0.0            0.0

[53 rows x 50 columns]

In [101]:
helmer_enc.mapping[0]['mapping']


where_town_0  where_town_1  ...  where_town_48  where_town_49
 1           -1.0          -1.0  ...           -1.0           -1.0
 2            1.0          -1.0  ...           -1.0           -1.0
 3            0.0           2.0  ...           -1.0           -1.0
 4            0.0           0.0  ...           -1.0           -1.0
 5            0.0           0.0  ...           -1.0           -1.0
 6            0.0           0.0  ...           -1.0           -1.0
 7            0.0           0.0  ...           -1.0           -1.0
 8            0.0           0.0  ...           -1.0           -1.0
 9            0.0           0.0  ...           -1.0           -1.0
 10           0.0           0.0  ...           -1.0           -1.0
 11           0.0           0.0  ...           -1.0           -1.0
 12           0.0           0.0  ...           -1.0           -1.0
 13           0.0           0.0  ...           -1.0           -1.0
 14           0.0           0.0  ...           -1.0           -1.0
 15           0.0           0.0  ...           -1.0           -1.0
 16           0.0           0.0  ...           -1.0           -1.0
 17           0.0           0.0  ...           -1.0           -1.0
 18           0.0           0.0  ...           -1.0           -1.0
 19           0.0           0.0  ...           -1.0           -1.0
 20           0.0           0.0  ...           -1.0           -1.0
 21           0.0           0.0  ...           -1.0           -1.0
 22           0.0           0.0  ...           -1.0           -1.0
 23           0.0           0.0  ...           -1.0           -1.0
 24           0.0           0.0  ...           -1.0           -1.0
 25           0.0           0.0  ...           -1.0           -1.0
 26           0.0           0.0  ...           -1.0           -1.0
 27           0.0           0.0  ...           -1.0           -1.0
 28           0.0           0.0  ...           -1.0           -1.0
 29           0.0           0.0  ...           -1.0           -1.0
 30           0.0           0.0  ...           -1.0           -1.0
 31           0.0           0.0  ...           -1.0           -1.0
 32           0.0           0.0  ...           -1.0           -1.0
 33           0.0           0.0  ...           -1.0           -1.0
 34           0.0           0.0  ...           -1.0           -1.0
 35           0.0           0.0  ...           -1.0           -1.0
 36           0.0           0.0  ...           -1.0           -1.0
 37           0.0           0.0  ...           -1.0           -1.0
 38           0.0           0.0  ...           -1.0           -1.0
 39           0.0           0.0  ...           -1.0           -1.0
 40           0.0           0.0  ...           -1.0           -1.0
 41           0.0           0.0  ...           -1.0           -1.0
 42           0.0           0.0  ...           -1.0           -1.0
 43           0.0           0.0  ...           -1.0           -1.0
 44           0.0           0.0  ...           -1.0           -1.0
 45           0.0           0.0  ...           -1.0           -1.0
 46           0.0           0.0  ...           -1.0           -1.0
 47           0.0           0.0  ...           -1.0           -1.0
 48           0.0           0.0  ...           -1.0           -1.0
 49           0.0           0.0  ...           -1.0           -1.0
 50           0.0           0.0  ...           49.0           -1.0
 51           0.0           0.0  ...            0.0           50.0
-1            0.0           0.0  ...            0.0            0.0
-2            0.0           0.0  ...            0.0            0.0

[53 rows x 50 columns]

Utilizando el ejemplo de prueba

In [102]:
helmer_enc = HelmertEncoder()
helmer_df = helmer_enc.fit_transform(df_test['City'])
helmer_df['Rent']=df_test['Rent']
helmer_df


/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



intercept  City_0  City_1  Rent
0          1    -1.0    -1.0  3999
1          1    -1.0    -1.0  4000
2          1    -1.0    -1.0  4001
3          1     1.0    -1.0  3499
4          1     1.0    -1.0  3500
5          1     1.0    -1.0  3501
6          1     0.0     2.0  2499
7          1     0.0     2.0  2500
8          1     0.0     2.0  2501

In [104]:
helmer_enc.mapping[0]['mapping']


City_0  City_1
 1    -1.0    -1.0
 2     1.0    -1.0
 3     0.0     2.0
-1     0.0     0.0
-2     0.0     0.0

In [105]:
model = linear_model.LinearRegression()
model.fit(helmer_df[['City_0','City_1']], helmer_df['Rent'])


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [106]:
print(f'Coefficients: {model.coef_}')
print(f'Intercept: {model.intercept_}')

Coefficients: [-250.         -416.66666667]
Intercept: 3333.3333333333335


In [107]:
print('Global mean: ', df_test['Rent'].mean())
print(df_test.groupby('City').mean())

Global mean:  3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


In [108]:
df_test[df_test['City']!='NYC']['Rent'].mean()

3250.0

#Supervised Encoders

In [109]:
col_cat = 'drinks'
col_num = 'essay_length'

In [110]:
from sklearn.preprocessing import LabelEncoder

In [111]:
from sklearn.preprocessing import LabelEncoder
label_enc = LabelEncoder()
df['Class_enc'] = label_enc.fit_transform(df['Class'])

In [112]:
label_enc.fit_transform(df[[col]])


/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:251: DataConversionWarning:

A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().



array([47,  5, 40, ..., 47, 40, 40])

## Target Encoder

In [115]:
from category_encoders import TargetEncoder
target_enc = TargetEncoder()


In [116]:
target_enc.fit_transform(df[col_cat], df['Class_enc'])


/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.189201
4      0.189201
6      0.189201
8      0.189201
10     0.147891
...         ...
59940  0.189201
59943  0.195429
59944  0.147891
59945  0.189201
59946  0.189201

[38809 rows x 1 columns]

In [117]:
target_enc.mapping

{'drinks': drinks
  1    0.189201
  2    0.147891
  3    0.170058
  4    0.195429
  5    0.179636
  6    0.137931
  7    0.171429
 -1    0.184674
 -2    0.184674
 dtype: float64}

In [118]:
target_enc = TargetEncoder()
target_enc.fit_transform(df[col_cat],df[col_num])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      3.136058
4      3.136058
6      3.136058
8      3.136058
10     3.145403
...         ...
59940  3.136058
59943  3.138155
59944  3.145403
59945  3.136058
59946  3.136058

[38809 rows x 1 columns]

In [119]:
target_enc.mapping

{'drinks': drinks
  1    3.136058
  2    3.145403
  3    3.164222
  4    3.138155
  5    2.963267
  6    2.915820
  7    3.072538
 -1    3.132119
 -2    3.132119
 dtype: float64}

## Leave-one-out Encoder

In [120]:
from category_encoders import LeaveOneOutEncoder
leave_enc = LeaveOneOutEncoder(sigma=0.05)


In [121]:
leave_enc.fit_transform(df[col_cat], df['Class_enc'])


drinks
1      0.167279
4      0.170307
6      0.187649
8      0.200296
10     0.133553
...         ...
59940  0.188762
59943  0.191911
59944  0.149245
59945  0.191099
59946  0.182342

[38809 rows x 1 columns]

In [122]:
leave_enc.mapping


{'drinks':                    sum  count
 desperately       36.0    210
 drinks_missing   217.0   1208
 not_at_all       319.0   2157
 often            667.0   3413
 rarely           677.0   3981
 socially        5207.0  27521
 very_often        44.0    319}

In [123]:
low_enc = LeaveOneOutEncoder()
low_enc.fit_transform(df[col_cat], df['essay_length'])

drinks
1      3.136048
4      3.136075
6      3.136049
8      3.136060
10     3.145293
...         ...
59940  3.136040
59943  3.138056
59944  3.145347
59945  3.136052
59946  3.136044

[38809 rows x 1 columns]

In [124]:
lou_enc.mapping

NameError: ignored

## M-estimate Encoder

In [125]:
from category_encoders import MEstimateEncoder
mest_enc = MEstimateEncoder(m=10)


In [126]:
mest_enc.fit_transform(df[col_cat], df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.189199
4      0.189199
6      0.189199
8      0.189199
10     0.148060
...         ...
59940  0.189199
59943  0.195398
59944  0.148060
59945  0.189199
59946  0.189199

[38809 rows x 1 columns]

In [127]:
mest_enc.mapping


{'drinks': drinks
  1    0.189199
  2    0.148060
  3    0.170094
  4    0.195398
  5    0.179677
  6    0.139352
  7    0.172031
 -1    0.184674
 -2    0.184674
 dtype: float64}

In [128]:
mlow_enc = LeaveOneOutEncoder()
mlow_enc.fit_transform(df[col_cat], df['essay_length'])


drinks
1      3.136048
4      3.136075
6      3.136049
8      3.136060
10     3.145293
...         ...
59940  3.136040
59943  3.138056
59944  3.145347
59945  3.136052
59946  3.136044

[38809 rows x 1 columns]

In [129]:
mlow_enc.mapping


{'drinks':                          sum  count
 desperately       645.232955    210
 drinks_missing   3579.626046   1208
 not_at_all       6784.635061   2157
 often           10710.522351   3413
 rarely          12596.767259   3981
 socially        86307.461805  27521
 very_often        930.146539    319}

## Weight of Evidence Encoder

In [130]:
from category_encoders import WOEEncoder 

In [131]:
woe_enc = WOEEncoder()
woe_enc.fit_transform(df[col_cat],df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.029719
4      0.029719
6      0.029719
8      0.029719
10    -0.263874
...         ...
59940  0.029719
59943  0.070806
59944 -0.263874
59945  0.029719
59946  0.029719

[38809 rows x 1 columns]

In [132]:
woe_enc.mapping

{'drinks': drinks
  1    0.029719
  2   -0.263874
  3   -0.099262
  4    0.070806
  5   -0.030446
  6   -0.328956
  7   -0.069086
 -1    0.000000
 -2    0.000000
 dtype: float64}